client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java

```diff
@@ -1525,9 +1524,13 @@ public void setupMetaServiceRef(RpcEndpointRef endpointRef) {
     driverRssMetaService = endpointRef;
   }
 
-  protected boolean mapperEnded(int shuffleId, int mapId, int attemptId) {
-    return mapperEndMap.containsKey(shuffleId)
-        && mapperEndMap.get(shuffleId).contains(Utils.makeMapKey(shuffleId, mapId, attemptId));
+  protected boolean mapperEnded(int shuffleId, int mapId) {
+    return (mapperEndMap.containsKey(shuffleId) && mapperEndMap.get(shuffleId).contains(mapId))
+        || stageEnded(shuffleId);
```

Comment:
how about putting `stageEnded(shuffleId)` at first?

Reply:
I think mapperEnd may be more likely.